src/hotspot/share/adlc/archDesc.cpp

```diff
@@ -684,6 +685,88 @@ bool ArchDesc::verify() {
   return true;
 }
 
+class MarkUsageFormClosure : public FormClosure {
+private:
+  ArchDesc* _ad;
+  std::unordered_set<Form*> *_visited;
+
+public:
+  MarkUsageFormClosure(ArchDesc* ad, std::unordered_set<Form*> *visit_map) {
+    _ad = ad;
+    _visited = visit_map;
+  }
+  virtual ~MarkUsageFormClosure() = default;
+
+  virtual void do_form(Form *form) {
+    if (_visited->find(form) == _visited->end()) {
+      _visited->insert(form);
+      form->forms_do(this);
+    }
+  }
+
+  virtual void do_form_by_name(const char* name) {
+    const Form* form = _ad->globalNames()[name];
+    if (form) {
+      do_form(const_cast<Form*>(form));
+      return;
+    }
+    RegisterForm* regs = _ad->get_registers();
+    if (regs->getRegClass(name)) {
+      do_form(regs->getRegClass(name));
+      return;
+    }
+  }
+};
+
+// check unused operands
+bool ArchDesc::check_usage() {
+  std::unordered_set<Form*> visited;
+  MarkUsageFormClosure callback(this, &visited);
+  _instructions.reset();
+  // iterate all instruction to mark used form
+  InstructForm* instr;
+  for ( ; (instr = (InstructForm*)_instructions.iter()) != nullptr; ) {
+    callback.do_form(instr);
+  }
+
+  // these forms are coded in OperandForm::is_user_name_for_sReg
+  callback.do_form_by_name("stackSlotI");
+  callback.do_form_by_name("stackSlotP");
+  callback.do_form_by_name("stackSlotD");
+  callback.do_form_by_name("stackSlotF");
+  callback.do_form_by_name("stackSlotL");
+
+  // special generic vector operands used in Matcher::pd_specialize_generic_vector_operand
```

Comment:
What is special about `stackSlot*` and `vec*` operands to process them explicitly here?

Reply:
I actually asked why `callback.do_form(instr)` calls in loop are not marking them as used? Why you need to call `do_form_by_name()` for them?

Reply:
For example, stackSlotP in aarch64 is not used. If not do_form_by_name, it will be reported as unused.

Reply:
Got it. Please, add comment describing this special case for vector and stack operators.

Reply:
Thanks for suggestion. I added the example in comments.

Reply:
Good. Please update comment for vectors too (add `only`):
```
// special generic vector operands only used in Matcher::pd_specialize_generic_vector_operand
```

I will submit our testing meanwhile.

Reply:
comments fixed